### DISCUSSION TITLE
Hibernation

### DISCUSSION THREAD

**Discussion by michael-hiegl** [linuxmint/discussions#112] [Ideas]
How about making hibernation to disc a default feature?

**Comment by claudiux** [comment]
https://askubuntu.com/questions/1258340/why-isnt-hibernation-available-by-default-in-ubuntu

**Reply by carsten88** [reply]
@claudiux
Thank you, so it IS possible, just takes some diskspace.

**Comment by carsten88** [comment]
Think about how much energy would be saved worldwide if this option was included and shipped as the default set option. This would be good news for the planet and all living being. Everyone I know just shuts the lid and leaves it on all night.

**Comment by cplittlfield** [comment]
I greatly prefer Hibernate to Suspend, given the price of electricity in my area.  Hibernating worked just fine for me on LM 21._x_, but I can't get it to work on LM 22.1.  Apparently a particular parameter was removed  by Ubuntu, and I haven't found where to restore it yet, even working from a solution from the Mint forums.

If you ask about hibernation on the forums, you get a large number of discouraging responses from the users who provide the best advice, on the grounds that it causes problems on many systems, which is why Ubuntu apparently stopped enabling hibernation by default.  However, for users like me, who can use the feature successfully when it is enabled, it would be a boon.

**Reply by carsten88** [reply]
O well thats a shame. Maybe LMDE has this feature?

**Comment by Heavygrass** [comment]
I use hibernation every single day w/o any issue.
I can surely believe it can cause an issue on some systems, but that's not a reason to disable it for everybody.
One should device for him/herself whether its worth to use it or not.
Its especially useful for a laptop.

**Comment by alosslessdev** [comment]
It's possible but doing so would require diverting somewhat from ubuntu which has disabled it

**Reply by Heavygrass** [reply]
Why does it matter?
All people are asking here is to enable GUI hibernate button by default.

**Reply by cplittlfield** [reply]
It's far more complicated than making a button available in a dialogue box.

For instance, how much RAM do you have?  Hibernating requires an amount of swap space at least the size of the RAM, plus an extra margin.   So a system with 32 GB of RAM would require at least 38 GB of swap space.  Yet I've seen complaints on the forums that even a mere 2 GB swap file is too large on a system with 32 GB of RAM.

In short, I don't think the devs are interested in reversing every change Ubuntu has made to prevent hibernation.  But who knows?  I might be wrong.

**Reply by Heavygrass** [reply]
> So a system with 32 GB of RAM would require at least 38 GB of swap space. Yet I've seen complaints on the forums that even a mere 2 GB swap file is too large on a system with 32 GB of RAM.

You're contradicting yourself here. You say "2 GB swap file is too large on a system with 32 GB of RAM.", but you need citing "at least 38 GB of swap space"?
32 GB +2GB=34GB, not 38 GB (add .1GB for safety).

> Hibernating requires an amount of swap space at least the size of the RAM

This is not always true. Depends on your RAM usage.
But again this is **irrelevant to Cinnamon** and whats being asked here.
Let the user handle the swap.

> in reversing every change Ubuntu has made to prevent hibernation.

What every change?
I have experience turning on hibernation on Ubuntu Cinnamon 24.04. The hardest part was getting back hibernation button in Cinnamon.

**Reply by cplittlfield** [reply]
At the very least, for hibernation to work, it is true that the size of the available swap space _must_ exceed the amount of RAM on the system.  Just how much extra swap is required might be debatable, but when the system hibernates, it writes the entire contents of RAM to swap.  And if there was already data in the swap, then . . . .

The AskUbuntu Web site has a discussion of hibernation that lists the amount of swap space required for different amounts of RAM.  So for example, to hibernate a system with 4 GB of RAM one needs 6-8 GB of swap space, 8 GB of RAM needs 11-16 GB of swap, 12 GB needs 15-24 GB, and so forth.  So a system with 32 GB of RAM needs 38-64 GB of swap to hibernate successfully.  

My point was that I am seeing a lot of complaints that "Why do I need a 2 GB swap file when I have 32 GB of RAM?  Why does it have to be so large?"  (As you may know, the Ubiquity installer automatically creates a swap file of that size.)  So if they react this badly to 2 GB of swap, what will their reaction be when the installer creates a 38 GB swap file?

Part of the problem is that former Windows users are used to seeing a dynamic swap file, which enlarges and shrinks, depending on need.  In Linux, swap space is fixed.

**Reply by claudiux** [reply]
I often use 1.5 times the amount of RAM for swap, when the RAM is greater than or equal to 8 GiB (2 times the amount of RAM when the RAM is less than 8 GiB). No problems with hibernation.

**Reply by alosslessdev** [reply]
> Part of the problem is that former Windows users are used to seeing a dynamic swap file, which enlarges and shrinks, depending on need. In Linux, swap space is fixed.

A package such as https://github.com/Tookmund/Swapspace could be used but then the kernel would need a patch to hibernate across several swapfiles. Maybe that swapspace code could be incorporated into the kernel too, but it would add technical debt. The right way to do things would be to remake the way the Linux kernel handles swap and implement dynamic swap in the kernel which no doubt would create pushback and would also need a lot of testing. It could be a kernel compile time option. Probably also add a limit in case a process is leaking memory.

**Reply by cplittlfield** [reply]
I'm not sure patching the kernel is necessary.  I have read that Linux can use any and all swap space on the system, whether partitions or files.  I can't find the passage I have in mind, but the swapon man page says
>        Calls to swapon normally occur in the system boot scripts making all swap devices
>        available, so that the paging and swapping activity is interleaved across several
>        devices and files.

**Reply by Heavygrass** [reply]
> At the very least, for hibernation to work, it is true that the size of the available swap space _must_ exceed the amount of RAM on the system.

Are you willing to make a cash bet on your statement?
I'm sure we could find a site to do that. How about 500$+ bet?

Are you claiming that if swapfile is less than the amount of RAM the hibernation will NEVER EVER work?

> So if they react this badly to 2 GB of swap, what will their reaction be when the installer creates a 38 GB swap file?

Why does installer has to create any swap file at all?
What does it have to do with Cinnamon specifically?

> Just how much extra swap is required might be debatable

So how much exactly you're claiming is the minimum extra swap?

**Reply by alosslessdev** [reply]
what about crating the swap file at hibernate time, only if there's enough space? Or, if the laptop is about to shut down, do it after the battery reaches a certain threshold? maybe even based on battery wear levels? The linux kernel should have a way to calculate how much swap is needed for hibernation, maybe it is current swap use + used ram that isn't available?

**Comment by cplittlfield** [comment]
I'm going only by what I've read on the Linux Mint and AskUbuntu forums.  As recently as LM 21.3, a swap file was required for writing the RAM image and restoring from it, and I am quoting the recommendations I've read.  The kernel documentation at www.kernel.org says the following:
>  Hibernation
> 
> This state (also referred to as Suspend-to-Disk or STD) offers the greatest energy savings and can be used even in the absence of low-level platform support for system suspend. However, it requires some low-level code for resuming the system to be present for the underlying CPU architecture.
> 
> Hibernation is significantly different from any of the system suspend variants. It takes three system state changes to put it into hibernation and two system state changes to resume it.
> 
> First, when hibernation is triggered, the kernel stops all system activity and creates a snapshot image of memory to be written into persistent storage. Next, the system goes into a state in which the snapshot image can be saved, the image is written out and finally the system goes into the target low-power state in which power is cut from almost all of its hardware components, including memory, except for a limited set of wakeup devices.
> 
> Once the snapshot image has been written out, the system may either enter a special low-power state (like ACPI S4), or it may simply power down itself. Powering down means minimum power draw and it allows this mechanism to work on any system. However, entering a special low-power state may allow additional means of system wakeup to be used (e.g. pressing a key on the keyboard or opening a laptop lid).

If anyone is saying that the contents of RAM do not need to be written to disk for hibernating, I'd like to see a reference for that.  If anyone is saying that the contents of RAM no longer need to be written to swap, then I'd like to see a reference for that, as well.  I'd like to get hibernation re-enabled on my LM 22.1 setup (it worked just fine on LM 21._x_), so if you were able to do that, please message me with instructions.

**Reply by Heavygrass** [reply]
> If anyone is saying that the contents of RAM do not need to be written to disk for hibernating, I'd like to see a reference for that. If anyone is saying that the contents of RAM no longer need to be written to swap, then I'd like to see a reference for that, as well

In my world the **one who makes a general statement MUST provide the proof** for the statement. You made general statements about hibernation, not me.
Given that you ignored every single of my questions I am taking it as you are not standing behind any of your generic claims.
You refused to even answer what exactly is your statement about swap file size is.

**Reply by cplittlfield** [reply]
[[GUIDE] How to hibernate to a swap file in Linux Mint 19.x](https://forums.linuxmint.com/viewtopic.php?t=284100) (topic still valid on LM 21.3)
[How to enable hibernation with swap partition on Linux Mint 19](https://forums.linuxmint.com/viewtopic.php?p=1495203#p1495203)
https://www.kernel.org/doc/Documentation/power/swsusp.txt
https://www.kernel.org/doc/Documentation/power/swsusp-and-swap-files.txt

### RHEL 8 table:
Amount of RAM in the system | Recommended swap space | Recommended for hibernation
-- | -- | --
≤ 2 GB | 2 times the amount of RAM | 3 times the amount of RAM
2 GB – 8 GB | Equal to the amount of RAM | 2 times the amount of RAM
8 GB – 64 GB | At least 4 GB | 1.5 times the amount of RAM
64+ GB | At least 4 GB | Hibernation not recommended

**Reply by Heavygrass** [reply]
What relevance does this have to my questions above to **you** specifically?
Once again, my questions are to **you** (not to RHEL, not to some online page etc).
Do you have any **personal** opinion on minimum swap size or not? (yes or no please)

**Reply by cplittlfield** [reply]
Oh, for heaven's sake!  You asked for sources of what I posted.  I gave them to you.  If you disagree with those sources, then refute them.  But don't try to change the goal posts, it won't work.  Do _you_ have opinions, and what are the sources _you_ base them on?

See, two can play that stupid game.

**Reply by Heavygrass** [reply]
I **never asked you for any sources**.
You made a generic statement, I asked you for **your** personal clarifications of your own statement. Period.
My questions to you are listed above (unchanged btw) and are very clear and specific and you keep avoiding every single one of my very specific questions again and again in various ways (for weeks already).
As of today you still have NOT answered a single of my questions above. Why?

**Reply by Heavygrass** [reply]
> But don't try to change the goal posts, it won't work.

What of my "goal posts" you are talking about?
Please provide a citation of my "goal posts".

**Reply by cplittlfield** [reply]
I refer you to your own words:
> In my world the one who makes a general statement MUST provide the proof for the statement. You made general statements about hibernation, not me.
> Given that you ignored every single of my questions I am taking it as you are not standing behind any of your generic claims.
You refused to even answer what exactly is your statement about swap file size is.

These are the sources for my statement that swapfile size must equal or exceed RAM size for hibernation to work.  You questioned it, I provided proof.  Now, stop bothering me.  I'm done here.

**Reply by Heavygrass** [reply]
> You questioned it

So where have you found a request for proof?
You still refuse to even answer what exactly your statement about minimum swap size is, which I asked for weeks ago.
There's zero sense to talk about a proof until there's a **clear** statement.

> . I'm done here.

Good finally, as I thought you don't stand behind yours statements. So no generic claims about the minimum swap size.

**Reply by Heavygrass** [reply]
> These are the sources for my statement that swapfile size must equal or exceed RAM size for hibernation to work.

Let me repeat my **weeks old** question for you again:
"Are you claiming that if swapfile is less than the amount of RAM the hibernation will NEVER EVER work?".
FYI this is a **YES/NO** question.
Just in case: its NOT asking for any source/proof etc.

**Comment by alosslessdev** [comment]
Today I tested my machine a Lenovo 15ach6h on debian 13 with KDE plasma and hibernation works out of the box. No configuration required. Granted I haven't even tried installing Nvidia drivers. So I'm convinced that a big barrier is Ubuntu and undoing their changes. That maybe explains why some people are able to hibernate and others aren't. However hibernation takes much longer than windows and it only works once per boot. After you try it once subsequent attempts fail to shut down the machine I guess due to nouveau driver problems.

**Reply by cplittlfield** [reply]
That's news to me.  On my HP with Intel graphics running the i915 driver, you could hibernate as many times as desired under 21._x_.  So you are probably right about it being something to do with the nouveau driver.
